compiler/src/dotty/tools/dotc/transform/ElimRepeated.scala

```diff
@@ -30,136 +30,208 @@ class ElimRepeated extends MiniPhase with InfoTransformer { thisPhase =>
-
-  /** If method overrides a Java varargs method, add a varargs bridge.
-   *  Also transform trees inside method annotation
+  /** If method overrides a Java varargs method or is annotated with @varargs, add a varargs bridge.
+   *  Also transform trees inside method annotation.
    */
-  override def transformDefDef(tree: DefDef)(implicit ctx: Context): Tree =
+  override def transformDefDef(tree: DefDef)(using ctx: Context): Tree =
     ctx.atPhase(thisPhase) {
-      if (tree.symbol.info.isVarArgsMethod && overridesJava(tree.symbol))
-        addVarArgsBridge(tree)
+      val sym = tree.symbol
+      val hasAnnotation = hasVarargsAnnotation(sym)
+      if hasRepeatedParams(sym) then
+        val isOverride = overridesJava(sym)
+        if isOverride || hasAnnotation || parentHasAnnotation(sym) then
+          // java varargs are more restrictive than scala's
+          // see https://github.com/scala/bug/issues/11714
+          if !isValidJavaVarArgs(sym.info) then
+            ctx.error("""To generate java-compatible varargs:
+                      |  - there must be a single repeated parameter
+                      |  - it must be the last argument in the last parameter list
+                      |""".stripMargin,
+              tree.sourcePos)
+            tree
+          else
+            addVarArgsBridge(tree, isOverride)
+        else
+          tree
       else
+        if hasAnnotation then
+          ctx.error("A method without repeated parameters cannot be annotated with @varargs", tree.sourcePos)
         tree
     }
 
+  /** Is there a repeated parameter in some parameter list? */
+  private def hasRepeatedParams(sym: Symbol)(using Context): Boolean =
+    sym.info.paramInfoss.flatten.exists(_.isRepeatedParam)
+
+  /** Is this the type of a method that has a repeated parameter type as
+   *  its last parameter in the last parameter list?
+   */
+  private def isValidJavaVarArgs(tp: Type)(using Context): Boolean = tp match
+    case mt: MethodType =>
+      val initp :+ lastp = mt.paramInfoss
+      initp.forall(_.forall(!_.isRepeatedParam)) &&
+      lastp.nonEmpty &&
+      lastp.init.forall(!_.isRepeatedParam) &&
+      lastp.last.isRepeatedParam
+    case pt: PolyType =>
+      isValidJavaVarArgs(pt.resultType)
+    case _ => false
+
+
   /** Add a Java varargs bridge
-   *  @param  ddef  the original method definition which is assumed to override
-   *                a Java varargs method JM up to this phase.
-   *  @return  a thicket consisting of `ddef` and a varargs bridge method
-   *           which overrides the Java varargs method JM from this phase on
-   *           and forwards to `ddef`.
+   *  @param ddef    the original method definition
+   *  @param addFlag the flag to add to the method symbol
+
+   *  @return a thicket consisting of `ddef` and a varargs bridge method
+   *          which forwards java varargs to `ddef`. It retains all the
+   *          flags of `ddef` except `Private`.
```

Comment:
Private flag is no longer removed

Reply:
I don't think so since a private method can't override anything any way.